doc/SONiC_YANG_Model_Guidelines.md

```diff
@@ -6,46 +6,153 @@
+
+Example :
+####  YANG
+```
+module sonic-acl {
+	container sonic-acl {
+		.....
+		.....
+	}
+}
+```
+
+
+### 2  It is best to categorize YANG Modules based on a networking components. For example, it is good to have separate modules for VLAN,  ACL, PORT and IP-ADDRESSES etc.
+```
+sonic-acl.yang
+sonic-interface.yang
+sonic-port.yang
+sonic-vlan.yang
+```
+
+### 3. Define namespace as "http://github.com/Azure/{model-name}" and define a shorter(preferably within 3-4 letters) namespace prefix. Use the prefix in XPath for referring to any node present in other model.
+
+Example :
+####  YANG
+```
+module sonic-acl {
+	namespace "http://github.com/Azure/sonic-acl";
+	prefix sacl;
+	.....
+	.....
+}
+```
+
+### 4. Use 'revision' to record revision history whenever YANG model is changed. Updating revision with appropriate details helps tracking the changes in the model.
+
+Example :
+
+#### YANG
+```
+module sonic-acl {
+	revision 2019-09-02 {
+		description
+			"Added must expression for ACL_RULE_LIST."; // Updated with new change details
+	}
+	
+	revision 2019-09-01 {
+		description
+			"Initial revision.";
+	}
+
+	container sonic-acl {
+		.....
+		.....
+	}
+}
+```
+
+### 5. Each primary section of ABNF.json (i.e a dictionary in ABNF.json) for example, VLAN, VLAN_MEMBER, INTERFACE  in ABNF.json will be mapped to a container in YANG model.
+
+Example: Table VLAN will translate to container VLAN.
 
 #### ABNF
 
 ```
-"INTERFACE": {
-        "Ethernet112|2a04:fxxx:40:a709::xxx/126": {
-            "scope": "global",
-            "family": "IPv6"
+"VLAN": {
+        "Vlan100": {
+             "vlanid": "107"
         }
  }
 ```
 will translate to:
 
 ####  YANG
 --
+```
+container VLAN {  //"VLAN" mapped to a container
+ list VLAN_LIST {
+   key name;
+   leaf vlanid {
+      type uint16;
+   }
+ }
+}
+```
+
+### 6. By default table is defined in CONFIG_DB, if needed use extension 'sonic-ext:db-name' for defining the table present in other Redis DB.
+
+####  YANG
+
 ```
 container INTERFACE {
+sonic-ext:db-name "APPL_DB"; //table defined in APPL_DB 
 .....
 .....
 }
 ```
 
-### 2. Each leaf in YANG module should have same name as corresponding key-fields in ABNF.json.
+### 7. The default separator used in table key  is "|". If it is different, use 'sonic-ext:key-delim {separator};' YANG extension.
+
+Example : 
+```
+container ACL_TABLE {
+	list ACL_TABLE_LIST {
+		sonic-ext:key-delim ":";
+		.....
+		.....
+	}
+}
+```
+
+### 8. Each leaf in YANG module should have same name (including their case) as corresponding key-fields in ABNF.json.
```

Comment:
There is no case here. All keys in redis are UPPER case only. And all fields are in lower case only.

So as for schema, we pick one case and have it uniform.
Seeing container & list names in lower case, I tend toward using lowercase everywhere.

The code that writes, when forming key-string from schema, it could make them upper case as part of coining key

Reply:
It is not correct that "And all fields are in lower case only.". Please take a look into  ACL_TABLE which has lower case fields and ACL_RULE which has upper case fields.

Reply:
Currently there is no set standard. So many daemons & tools, convert what they read into one case (mostly upper) and handle it. 

As part of this work, we could bring in an uniformity. Plus, a single way would make writing schema simpler and hence less error prone,

Here is my suggestion, which we need to float it out and get an agreement, from SONiC experts.
1) Have it all lower case in YANG schema
2) The DB layer, when it forms key / field name, convert all to upper.

SONiC YANG -- all lower
Internal: redis, config-db.json stay upper. The reason: Many daemons & tools already handle it as upper, so be in sync.

Reply:
I believe this is a separate discussion and has to be done carefully as it may break functionality in many areas like Click CLI, Application etc. Also, just want to make sure that, it is not only about field name, but also its value like (enum name etc.) are case sensitive today.

Let us proceed with "(including their case)" for now. We can easily remove it from guideline, once 'case' issue is sorted out across SONiC.